lib/_http_agent.js

```diff
@@ -392,13 +415,33 @@ Agent.prototype.removeSocket = function removeSocket(s, options) {
         // Don't leak
         if (sockets[name].length === 0)
           delete sockets[name];
+        this.totalSocketCount--;
       }
     }
   }
 
+  let req;
   if (this.requests[name] && this.requests[name].length) {
     debug('removeSocket, have a request, make a socket');
-    const req = this.requests[name][0];
+    req = this.requests[name][0];
+  } else {
+    // TODO(rickyes): this logic will not be FIFO across origins.
+    // There might be older requests in a different origin, but
+    // if the origin which releases the socket has pending requests
+    // that will be prioritized.
+    for (const prop in this.requests) {
+      // Check whether this specific origin is already at maxSockets
+      if (this.sockets[prop] && this.sockets[prop].length) break;
+      debug('removeSocket, have a request with different origin,' +
+        ' make a socket');
+      req = this.requests[prop][0];
```

Comment:
This doesn't actually remove the request from the freelist... the same request might/will get executed twice

Reply:
I don't think that's going to happen.

The free event will be emitted when the handle is successfully created: 
See https://github.com/nodejs/node/pull/33617/files#diff-5f7fb0850412c6be189faeddea6c5359R450

The delete request was done in the free event:
See https://github.com/nodejs/node/pull/33617/files#diff-5f7fb0850412c6be189faeddea6c5359R125-R132